packages/gatsby/src/utils/source-nodes.ts

```diff
@@ -70,19 +74,39 @@ function getStaleNodes(state: IGatsbyState, nodes: Array<Node>): Array<Node> {
       )
     }
 
-    return !state.nodesTouched.has(rootNode.id)
-  })
+    if (!state.nodesTouched.has(rootNode.id)) {
+      staleNodes.push(node)
+    }
+  }
+
+  const noNodesPlugins = state.flattenedPlugins
+    .filter(
+      plugin =>
+        // "Can generate nodes"
+        plugin.nodeAPIs.includes(`sourceNodes`) &&
+        // "Has not generated nodes"
+        !nodeOwnerSet.has(plugin.name)
+    )
+    .map(plugin => plugin.name)
```

Comment:
plugins aren't a big array but while you're optimizing shouldn't this be a forEach loop too so we don't have to loop over it twice?